src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -0,0 +1,231 @@
+package mesosphere.marathon
+package core.election
+
+import akka.actor.Cancellable
+import akka.stream.scaladsl.{ Source, SourceQueueWithComplete }
+import akka.stream.OverflowStrategy
+import com.typesafe.scalalogging.StrictLogging
+import java.util
+import java.util.Collections
+import java.util.concurrent.{ ExecutorService, TimeUnit }
+import mesosphere.marathon.metrics.{ Metrics, ServiceMetric, Timer }
+import mesosphere.marathon.stream.EnrichedFlow
+import mesosphere.marathon.util.LifeCycledCloseableLike
+import org.apache.curator.framework.CuratorFramework
+import org.apache.curator.framework.api.{ ACLProvider, CuratorWatcher, UnhandledErrorListener }
+import org.apache.curator.framework.imps.CuratorFrameworkState
+import org.apache.curator.framework.recipes.leader.LeaderLatch
+import org.apache.curator.framework.{ AuthInfo, CuratorFrameworkFactory }
+import org.apache.curator.retry.ExponentialBackoffRetry
+import org.apache.zookeeper.{ KeeperException, WatchedEvent, ZooDefs }
+import org.apache.zookeeper.data.ACL
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContext
+import scala.util.Try
+
+object CuratorElectionStream extends StrictLogging {
+  /**
+    * Connects to Zookeeper and offers leadership; monitors leader state. Watches for leadership changes (leader
+    * changed, was elected leader, lost leadership), and emits events accordingly.
+    *
+    * Materialized cancellable is used to abdicate leadership; which will do so followed by a closing of the stream.
+    */
+  def apply(
+    clientCloseable: LifeCycledCloseableLike[CuratorFramework],
+    zooKeeperLeaderPath: String,
+    zooKeeperConnectionTimeout: FiniteDuration,
+    hostPort: String,
+    singleThreadExecutor: ExecutorService): Source[LeadershipState, Cancellable] = {
+    val singleThreadEC = ExecutionContext.fromExecutor(singleThreadExecutor)
+    Source.queue[LeadershipState](16, OverflowStrategy.dropHead)
```

Comment:
Could we guarantee that only one event at a time can be processed?

Reply:
This is the idea with streams